text/3414-deprecate-spooky-dropck-at-a-distance.md

```diff
@@ -0,0 +1,362 @@
+    }
+}
+
+fn to_pd<T>(_: T) -> PhantomData<T> {
+    PhantomData
+}
+
+pub fn foo() {
+    let mut x;
+    {
+        let s = String::from("temporary");
+        let p = PrintOnDrop(&s);
+        x = (to_pd(p), ());
+    }
+}
+```
+
+`PhantomData`'s dropck behaviour is only checked if the type (in this case, the
+tuple) marks `needs_drop`, which is confusing. Unrelated tuple elements
+shouldn't affect `PhantomData` behaviour, but the above example shows that they
+do. This RFC makes it so they don't.
+
+Likewise, `[T; 0]` produces the same effect: it is not `needs_drop`, but adds
+outlive requirements, thus also exhibiting "spooky-dropck-at-a-distance".
+
+Simply defining every non-`needs_drop` type as being pure w.r.t. drop would,
+however, break `#[may_dangle]`, so we need to adapt it.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Changes to dropck (stable)
+
+Types like `fn(T)`, `ManuallyDrop<T>`, `PhantomData<T>`, `[T; 0]` and `&'a T`,
+that don't need to be dropped, place no liveness requirements on `'a` or `T`
+when they go out of scope - even if `T` would otherwise introduce liveness
+requirements.
+
+In other words, this compiles:
+
+```rust
+fn main() {
+    let mut x;
+    {
+        x = &String::new();
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+But a type which does need to be dropped, introduces liveness requirements:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x;
+    {
+        x = PrintOnDrop(&*String::new());
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+    // ERROR: temporary may not live long enough
+}
+```
+
+Unless it's in one of the above types:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x;
+    {
+        x = &PrintOnDrop(&*String::new());
+        // PrintOnDrop implicitly dropped here
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+As a special case, some built-in types, like `Vec`, `Box`, `Rc`, `Arc`,
+`HashMap`, among others, despite needing to be dropped, do not place liveness
+requirements unless `T` demands liveness requirements. This is okay:
+
+```rust
+fn main() {
+    let mut x = vec![];
+    {
+        x.push(&String::new());
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+But this isn't:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x = vec![];
+    {
+        x.push(PrintOnDrop(&*String::new()));
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
```

Comment:
```suggestion
    // The vector `x` is implicitly dropped here, as are any `PrintOnDrop<'_>` values contained within. Because `PrintOnDrop` implements `Drop`, the compiler assumes that these destructors may access the `&str` contained in their fields. These strings pointed to the dropped stack slot for `s`, so that is an error.
```